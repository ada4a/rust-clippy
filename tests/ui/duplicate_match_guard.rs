#![allow(clippy::needless_return, clippy::needless_else)]
#![warn(clippy::duplicate_match_guard)]

fn main() {
    let mut a = 5;
    let b = 4;
    let mut v: Vec<u32> = vec![];

    match 0u32 {
        0 if true => {
            if true {
                //~^ duplicate_match_guard
                return;
            }
        },
        0 if a > b => {
            if a > b {
                //~^ duplicate_match_guard
                return;
            }
        },
        // not _identical_, but the meaning is the same
        0 if a > b => {
            if b < a {
                //~^ duplicate_match_guard
                return;
            }
        },
        // a bit more complicated
        0 if a > 0 && b > 0 => {
            if a > 0 && b > 0 {
                //~^ duplicate_match_guard
                return;
            }
        },

        // no warnings
        0 if true => {
            if false {
                return;
            }
        },
        // has side effects, so don't lint
        0 if v.pop().is_some() => {
            if v.pop().is_some() {
                return;
            }
        },
        // there's _something_ in the else branch
        0 if a > b => {
            if a > b {
                a += b;
            } else {
            }
        },
        // body has statements before the if
        0 if a > b => {
            a += b;
            if a > b {
                return;
            }
        },
        _ => todo!(),
    };
}
