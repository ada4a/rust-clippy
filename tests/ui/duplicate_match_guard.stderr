error: condition duplicates match guard
  --> tests/ui/duplicate_match_guard.rs:11:16
   |
LL |             if true {
   |                ^^^^
   |
   = note: `-D clippy::duplicate-match-guard` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::duplicate_match_guard)]`

error: condition duplicates match guard
  --> tests/ui/duplicate_match_guard.rs:17:16
   |
LL |             if a > b {
   |                ^^^^^

error: condition duplicates match guard
  --> tests/ui/duplicate_match_guard.rs:24:16
   |
LL |             if b < a {
   |                ^^^^^

error: condition duplicates match guard
  --> tests/ui/duplicate_match_guard.rs:31:16
   |
LL |             if a > 0 && b > 0 {
   |                ^^^^^^^^^^^^^^

error: aborting due to 4 previous errors

